feat(report) profit and loss report by month

jeremielodi opened the pull request:
@jniles I'm working on #2799, try to have a look if I'm  on a good way

jeremielodi committed:
feat(report) profit and loss report by month

jniles commented on client/src/modules/reports/generate/income_expense_by_month/income_expense_by_month.config.js:
```diff
@@ -0,0 +1,65 @@
+
+function IncomeExpenseByMonthConfigController($sce, Notify, SavedReports, AppCache, reportData, $state) {
+  const vm = this;
+  const cache = new AppCache('configure_income_expenseBy_month');
```
For consistency, use `configure_income_expense_by_month`, since that is the URL.

jniles commented on server/controllers/finance/reports/income_expense_by_month/index.js:
```diff
@@ -0,0 +1,304 @@
+  Promise.all([
+    getPeriodByNumberAndFiscalId(parseInt(periodNumber, 10) - 1, fiscalYearId),
+    getPeriodById(periodId),
+    getPeriodByNumberAndFiscalId(parseInt(periodNumber, 10) + 1, fiscalYearId),
```
This should be the current period (Period N), the previous period (Period N -1), and the period before that one (Period N - 2).

jeremielodi replied:
This means the report compares the last tree monts' balanace?

jniles replied:
Yes sir!

jniles commented on server/controllers/finance/reports/income_expense_by_month/index.js:
```diff
@@ -0,0 +1,304 @@
+  const tree = new Tree(dataset);
+
+  const properties = [
+    'balance', 'previousBalance', 'difference', 'variance',
```
You will need to add `nextBalance` to this list to compute totals on that column.

jniles commented:
Code looks pretty good.  If a user chooses "January", an error gets thrown.  Can you see what you can do to prevent this?

Also, the changes from the last review should be implemented.

jeremielodi commented:
BHIMA itself handles a message when there is no data found in the database

jeremielodi committed:
fixing calculation error

jeremielodi committed:
fixing report decription

jeremielodi commented:
@mbayopanda, you can, if possible,  review this PR.

jeremielodi commented:
@mbayopanda this PR is ready for the review

mbayopanda commented on server/controllers/finance/reports/income_expense_by_month/index.js:
```diff
@@ -0,0 +1,302 @@
+    .then(([firstBalances, secondBalances, thirdBalances]) => {
+
+      const dataset = combineIntoSingleDataset(secondBalances, firstBalances, thirdBalances);
+      // console.log(dataset);
```
this comment must be removed

mbayopanda commented on server/controllers/finance/reports/income_expense_by_month/report.handlebars:
```diff
@@ -0,0 +1,90 @@
+
+      <h5 style="margin:15px;" class="text-center">
+        {{date thirdPeriod.start_date "MMMM"}} ({{ fiscalYear.label }})
+      </h5>
```
Could you use `text-capitalize` class.

```html
<h5 style="margin:15px;" class="text-center">
    <span class="text-capitalize">{{date thirdPeriod.start_date "MMMM"}}</span> ({{ fiscalYear.label }})
</h5>
```

jeremielodi commented:
@mbayopanda This PR is ready for a last review

jniles committed:
chore(reports): clean up income & expense by month

This commit updates the income and expense report by month to meet all
the standards set out in the code reviews:
 1. Variables have been updated to be called "first", "second", and
 "third" instead of "previous", "current", and "next".
 2. The totals are now bold, and colspans corrected.
 3. A summary box appears at the bottom of the report with the overall
 difference between profit and loss.
 4. Errors in date ranges are caught early, before hitting the database.
 5. Translations have been corrected, and the file name is the correct
 report name.
 6. Variance code is removed since it was unused.

This should round out all the pull request comments.

jniles commented:
@jeremielodi, please complete this report with the corrections that AG Neron proposed.

jeremielodi committed:
add avarage balance column

jniles committed:
Merge remote-tracking branch 'upstream/master' into profit_and_lot_by_month

jniles committed:
chore: add unit/report values to migrate.sql

This commit prepares the migration script for this report.

jniles committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jniles commented:
bors r+

jniles commented:
Seems like a fluke.

bors r+
